File: 线性表/linked-stack.c
#include <stdio.h>
#include <stdlib.h>
#define ElementType int

//堆栈的链式存储实现

typedef struct Node
{
	ElementType data;
	struct Node *next;
}Stack;

Stack *Create()
{
	Stack *p;
	p=(Stack *)malloc(sizeof(Stack));
	p->next=NULL;
	return p;
}

bool IsEmpty(Stack *p)
{
	if(p->next==NULL) return true;
	else return false;
}

void push(ElementType x,Stack *p)
{
	Stack *temp;
	temp=(Stack *)malloc(sizeof(Stack));
	temp->data=x;
	temp->next=p->next;
	p->next=temp;
}

ElementType pop(Stack *p)
{
	Stack *temp;
	ElementType x;
	if(p->next=NULL)
	{
		printf("stack is empty\n");
		return NULL;
	}
	temp=p->next;
	p->next=temp->next;
	x=temp->data;
	free(temp);
	return x;
}

File: data_structure/stack/linked_stack.py
"""
用链表实现栈
操作：
push():O(1)
pop():O(1)
"""


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedStack:
    def __init__(self):
        self.top = None

    def push(self, item):
        new_node = Node(item)
        if self.top == None:
            self.top = new_node
        else:
            new_node.next = self.top
            self.top = new_node

    def pop(self):
        if self.top == None:
            return None
        temp = self.top
        self.top = self.top.next
        return temp.data

    def is_empty(self):
        if self.top == None:
            return True
        return False

    def first(self):
        if self.top == None:
            return None
        
        return self.top.data

    def __str__(self):
        temp = self.top
        res = ''
        while temp:
            res += temp.data
            temp = temp.next
        return res[::-1]

if __name__ == "__main__":
    stack = LinkedStack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    print(stack.pop())
    print(stack.pop())
    print(stack.pop())
